### PR TITLE
Implementation of Soft Delete User and Update User

### PR DESCRIPTION
**Pull request by Rapter1990**
What I have done so far is focused on all these processes shown below.
1 ) Implementation of deleting user softly by the specified ID
2 ) Implementation of updating user by the specified ID

### CODE REVIEW

File: src/main/java/com/ays/backend/user/service/UserService.java [hunk 1]
```diff
@@ -41,4 +42,20 @@ public interface UserService {
...
+    UserDTO deleteSoftUserById(Long id);
+
+
+    /**
+     * Update User by User ID
+     * @param id the given User ID
+     * @param updateUserRequest the given UpdateUserRequest object
+     * @return userDto
+     */
+    UserDTO updateUserById(Long id, UpdateUserRequest updateUserRequest);
```
Comment (skayikci) [inline src/main/java/com/ays/backend/user/service/UserService.java:60]: Why don't you get the user id in the update user request object?

File: src/main/java/com/ays/backend/user/service/impl/UserServiceImpl.java [hunk 2]
```diff
@@ -86,5 +87,40 @@ public UserDTO getUserById(Long id) {
...
+        return userMapper.mapUsertoUserDTO(deactivedUser);
+    }
+
+    @Override
+    public UserDTO updateUserById(Long id, UpdateUserRequest updateUserRequest) {
+
+        User user = userRepository.findById(id)
+                .orElseThrow(() -> new UserNotFoundException("User with id " + id + " not found"));
+
+        user = User.builder()
```
Comment (skayikci) [inline src/main/java/com/ays/backend/user/service/impl/UserServiceImpl.java:108]: Please use MapStruct or something similar. And for the performance, please keep an eye for the usages.

File: src/main/java/com/ays/backend/user/controller/UserController.java [hunk 3]
```diff
@@ -66,4 +67,29 @@ public ResponseEntity<Page<UserDTO>> getUsers(PaginationRequest paginationReques
...
+
+    /**
+     * This endpoint returns a UserDTO object by updating the user with the specified ID.
+     *
+     * @param id A Long representing the ID of the user to retrieve (required).
+     * @return A ResponseEntity containing a UserDTO object after implementing the process of updating user
+     *         with the specified ID by and the HTTP status code (200 OK).
+     */
+    @PutMapping("/{id}")
+    public ResponseEntity<UserDTO> updateUserById(@PathVariable Long id, @RequestBody UpdateUserRequest updateUserRequest) {
```
[the remark this answers is not in the document]
Comment (Rapter1990) [reply src/main/java/com/ays/backend/user/controller/UserController.java]: I revised the method only getting UpdateUserRequest

File: src/main/java/com/ays/backend/user/service/impl/UserServiceImpl.java (hunk 2 above)
Comment (Rapter1990) [reply src/main/java/com/ays/backend/user/service/impl/UserServiceImpl.java]: I just revised the code block through MapStruct.

File: src/main/java/com/ays/backend/mapper/UserMapper.java [hunk 4]
```diff
@@ -1,13 +1,23 @@
...
 import com.ays.backend.user.model.entities.User;
 import com.ays.backend.user.service.dto.UserDTO;
+import org.mapstruct.AfterMapping;
 import org.mapstruct.Mapper;
+import org.mapstruct.Mapping;
+import org.mapstruct.MappingTarget;
+
+import java.time.LocalDateTime;
 
 @Mapper(componentModel = "spring")
```
Comment (skayikci) [inline src/main/java/com/ays/backend/mapper/UserMapper.java:13]: please change to:
@Mapper(componentModel = "spring", nullValuePropertyMappingStrategy = NullValuePropertyMappingStrategy.IGNORE)

File: src/main/java/com/ays/backend/user/service/impl/UserServiceImpl.java [hunk 5]
```diff
@@ -86,5 +87,32 @@ public UserDTO getUserById(Long id) {
...
+
+        user.setStatus(UserStatus.PASSIVE);
+
+        var deactivedUser =userRepository.save(user);
+
+        return userMapper.mapUsertoUserDTO(deactivedUser);
+    }
+
+    @Override
+    public UserDTO updateUserById(UpdateUserRequest updateUserRequest) {
```
Comment (skayikci) [inline src/main/java/com/ays/backend/user/service/impl/UserServiceImpl.java:103]: Please add    @Transactional

File: src/test/java/com/ays/backend/user/service/UserServiceTest.java [hunk 6]
```diff
@@ -149,7 +153,105 @@ public void shouldThrowUserNotFoundExceptionWhenUserNotExists() {
...
+
+        UserDTO deactivatedUser = userService.deleteSoftUserById(id);
+
+        // then
+        assertEquals(deactivatedUser.getUserStatus().ordinal(), userDTOInfoWithPassiveStatus.getUserStatus().ordinal());
+
+    }
+
+    @Test
+    public void shouldSoftDeleteUserByIdInvalidId() {
```
Comment (skayikci) [inline src/test/java/com/ays/backend/user/service/UserServiceTest.java:184]:
```suggestion
    public void shouldNotSoftDeleteUserWhenIdInvalidId() {
```

File: src/test/java/com/ays/backend/user/service/UserServiceTest.java [hunk 7]
```diff
@@ -149,7 +153,105 @@ public void shouldThrowUserNotFoundExceptionWhenUserNotExists() {
...
+        // then
+        assertEquals(updatedDTO.getUsername(), updatedUserDTO.getUsername());
+        assertEquals(updatedDTO.getFirstName(), updatedUserDTO.getFirstName());
+        assertEquals(updatedDTO.getLastName(), updatedUserDTO.getLastName());
+        assertEquals(updatedDTO.getUserRole().ordinal(), updatedUserDTO.getUserRole().ordinal());
+        assertEquals(updatedDTO.getUserStatus().ordinal(), updatedUserDTO.getUserStatus().ordinal());
+    }
+
+    @Test
+    public void shouldUpdateUserByIdInvalidId() {
```
Comment (skayikci) [inline src/test/java/com/ays/backend/user/service/UserServiceTest.java:235]:
```suggestion
    public void shouldNotUpdateUserWhenIdInvalidId() {
```

File: src/test/java/com/ays/backend/user/service/UserServiceTest.java [hunk 8]
```diff
@@ -149,7 +153,105 @@ public void shouldThrowUserNotFoundExceptionWhenUserNotExists() {
...
         UserNotFoundException actual = assertThrows(UserNotFoundException.class, () -> userService.getUserById(-1L));
 
+        // then
+        assertEquals(expectedError.getMessage(), actual.getMessage());
+
+    }
+
+
+    @Test
+    public void shouldSoftDeleteUserById() {
```
Comment (skayikci) [inline src/test/java/com/ays/backend/user/service/UserServiceTest.java:163]:
```suggestion
    void shouldSoftDeleteUserById() {
```

File: src/test/java/com/ays/backend/user/service/UserServiceTest.java [hunk 9]
```diff
@@ -149,7 +153,105 @@ public void shouldThrowUserNotFoundExceptionWhenUserNotExists() {
...
+        when(userRepository.findById(id)).thenReturn(Optional.empty());
+
+        UserNotFoundException actual = assertThrows(UserNotFoundException.class, () -> userService.deleteSoftUserById(id));
+
+        // then
         assertEquals(expectedError.getMessage(), actual.getMessage());
+    }
+
+    @Test
+    public void shouldUpdateUserById() {
```
Comment (skayikci) [inline src/test/java/com/ays/backend/user/service/UserServiceTest.java:200]:
```suggestion
    void shouldUpdateUserById() {
```

File: src/test/java/com/ays/backend/user/controller/UserControllerTest.java [hunk 10]
```diff
@@ -98,13 +102,69 @@ void shouldReturnGetUserDtoById() throws Exception {
...
+        mockMvc.perform(get(USER_CONTROLLER_BASEURL + "/{id}", id)
                     .contentType("application/json"))
                 .andExpect(status().isOk())
                 .andExpect(jsonPath("$.username").value("username 1"))
                 .andExpect(jsonPath("$.firstName").value("firstname 1"))
                 .andExpect(jsonPath("$.lastName").value("lastname 1"));
     }
 
+    @Test
+    void testDeleteSoftUserById() throws Exception {
```
Comment (skayikci) [inline src/test/java/com/ays/backend/user/controller/UserControllerTest.java:115]:
```suggestion
    void shouldSoftDeleteUserById() throws Exception {
```

File: src/test/java/com/ays/backend/user/controller/UserControllerTest.java [hunk 11]
```diff
@@ -98,13 +102,69 @@ void shouldReturnGetUserDtoById() throws Exception {
...
+                .contentType("application/json"))
+                .andExpect(status().isOk())
+                .andExpect(jsonPath("$.username", is(deletedSoftUser.getUsername())))
+                .andExpect(jsonPath("$.firstName", is(deletedSoftUser.getFirstName())))
+                .andExpect(jsonPath("$.lastName", is(deletedSoftUser.getLastName())))
+                .andExpect(jsonPath("$.userStatus", is(deletedSoftUser.getUserStatus().name())));
+    }
+
+    @Test
+    void testUpdateUserById() throws Exception {
```
[the remark this answers is not in the document]
Comment (Rapter1990) [reply src/test/java/com/ays/backend/user/controller/UserControllerTest.java:140]: I just revised the name of method

File: src/test/java/com/ays/backend/user/service/UserServiceTest.java (hunk 9 above)
Comment (Rapter1990) [reply src/test/java/com/ays/backend/user/service/UserServiceTest.java:200]: I just revised the name of method

File: src/test/java/com/ays/backend/user/service/UserServiceTest.java (hunk 8 above)
Comment (Rapter1990) [reply src/test/java/com/ays/backend/user/service/UserServiceTest.java:163]: I just revised the name of method

File: src/test/java/com/ays/backend/user/service/UserServiceTest.java (hunk 7 above)
Comment (Rapter1990) [reply src/test/java/com/ays/backend/user/service/UserServiceTest.java:235]: I just revised the name of method

File: src/test/java/com/ays/backend/user/service/UserServiceTest.java (hunk 6 above)
Comment (Rapter1990) [reply src/test/java/com/ays/backend/user/service/UserServiceTest.java:184]: I just revised the name of method

File: src/main/java/com/ays/backend/user/service/impl/UserServiceImpl.java (hunk 5 above)
Comment (Rapter1990) [reply src/main/java/com/ays/backend/user/service/impl/UserServiceImpl.java:103]: I add @Transactional above updateUserById method

**Comment by skayikci** [comment]
Please remove the commented out code in UserDto.

**Comment by skayikci** [comment]
@Rapter1990 do you think we should return passive users? Since when we delete them, we set the status to passive, and we can still get the user by passing his/her id.

**Review by skayikci — APPROVED** [review]

**Pull request merged**